File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        Stack<Integer> st = new Stack<Integer>();
        ListNode curr = head;
        while (curr != null)
        {
            st.push(curr.val);
            curr = curr.next;
        }
        ListNode cur = head;
        while (cur != null)
        {
            if (cur.val == st.peek())
            {
                st.pop();
                cur = cur.next;
            }
            else
            {
                return false;
            }
        }
        if (st.isEmpty() == true)
        {
            return true;
        }
        return false;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        curr = dcurr = head
        while dcurr.next and dcurr.next.next:
            curr = curr.next
            dcurr = dcurr.next.next
        prev = None
        ncurr = curr.next
        while ncurr:
            temp = ncurr.next
            ncurr.next = prev
            prev = ncurr
            ncurr = temp
        mcurr = head
        while mcurr and prev:
            if mcurr.val != prev.val:
                return False
            mcurr = mcurr.next
            prev = prev.next
        return True